From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Maddock <samuel.maddock@gmail.com>
Date: Tue, 20 Oct 2026 18:54:12 +0000
Subject: feat: add PreloadRealmGlobalScope

Electron runs service worker preload scripts in a context of their own on
the worker's thread. That context used to be a Blink ShadowRealmGlobalScope,
which exists for the TC39 ShadowRealm proposal. The proposal has stalled, so
Chromium may remove its implementation.

PreloadRealmGlobalScope is a global scope for that context which does not
depend on any ShadowRealm code. It is exposed the same interfaces as a
ShadowRealm, which are the ones the specs mark [Exposed=*], and defers to
the context that created it for everything else. Unlike a ShadowRealm, it
reports uncaught errors and console messages through that context instead
of treating them as unreachable.

Electron creates the context itself, in
shell/renderer/preload_realm_context.cc. This patch is specific to Electron
and is not intended to be upstreamed.

diff --git a/third_party/blink/renderer/bindings/core/v8/v8_initializer.cc b/third_party/blink/renderer/bindings/core/v8/v8_initializer.cc
index 41a5db4e019084d47763a43bd548b29faa070660..3e62033228307ef2474c91e250db08cd72934ad6 100644
--- a/third_party/blink/renderer/bindings/core/v8/v8_initializer.cc
+++ b/third_party/blink/renderer/bindings/core/v8/v8_initializer.cc
@@ -78,6 +78,7 @@
 #include "third_party/blink/renderer/core/inspector/inspector_trace_events.h"
 #include "third_party/blink/renderer/core/inspector/main_thread_debugger.h"
 #include "third_party/blink/renderer/core/page/page.h"
+#include "third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.h"
 #include "third_party/blink/renderer/core/script/modulator.h"
 #include "third_party/blink/renderer/core/shadow_realm/shadow_realm_global_scope.h"
 #include "third_party/blink/renderer/core/timing/time_clamper.h"
@@ -448,6 +449,11 @@ static void PromiseRejectHandlerInWorker(v8::PromiseRejectMessage data) {
   ExecutionContext* execution_context = ExecutionContext::From(script_state);
   if (!execution_context)
     return;
+  // An Electron preload realm has no script controller of its own.
+  if (auto* preload_realm =
+          DynamicTo<PreloadRealmGlobalScope>(execution_context)) {
+    execution_context = preload_realm->GetInitiatorExecutionContext();
+  }
 
   ExecutionContext* root_worker_context =
       execution_context->IsShadowRealmGlobalScope()
diff --git a/third_party/blink/renderer/bindings/generated_in_core.gni b/third_party/blink/renderer/bindings/generated_in_core.gni
index 788f261f96ef65df2bc7fef0fa2a793da22e44ca..7b8efe1bc88c18e3ccf6e9a663c6670f46e8bea6 100644
--- a/third_party/blink/renderer/bindings/generated_in_core.gni
+++ b/third_party/blink/renderer/bindings/generated_in_core.gni
@@ -1530,6 +1530,8 @@ generated_interface_sources_in_core = [
   "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_preference_manager.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_preference_object.cc",
   "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_preference_object.h",
+  "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_preload_realm_global_scope.cc",
+  "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_preload_realm_global_scope.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_progress_event.cc",
   "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_progress_event.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/core/v8/v8_promise_rejection_event.cc",
diff --git a/third_party/blink/renderer/bindings/generated_in_modules.gni b/third_party/blink/renderer/bindings/generated_in_modules.gni
index b56e9a96ebd3d598e53848fa955d02ac45628900..4f4d8a5ec938f7b4ccd336b64221d62017fb83b4 100644
--- a/third_party/blink/renderer/bindings/generated_in_modules.gni
+++ b/third_party/blink/renderer/bindings/generated_in_modules.gni
@@ -2793,6 +2793,8 @@ generated_interface_sources_in_modules = [
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_plugin.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_plugin_array.cc",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_plugin_array.h",
+  "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_preload_realm_global_scope.cc",
+  "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_preload_realm_global_scope.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_presentation.cc",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_presentation.h",
   "$root_gen_dir/third_party/blink/renderer/bindings/modules/v8/v8_presentation_availability.cc",
diff --git a/third_party/blink/renderer/bindings/idl_in_core.gni b/third_party/blink/renderer/bindings/idl_in_core.gni
index f8d8e7180904bdf3a330483b8e6e7c3bf5dde439..9f495478b3a5164db9f113041a8e4e38ffb1d0fd 100644
--- a/third_party/blink/renderer/bindings/idl_in_core.gni
+++ b/third_party/blink/renderer/bindings/idl_in_core.gni
@@ -828,6 +828,7 @@ static_idl_files_in_core = [
   "//third_party/blink/renderer/core/preferences/preference_manager.idl",
   "//third_party/blink/renderer/core/preferences/preference_object.idl",
   "//third_party/blink/renderer/core/preferences/navigator_preferences.idl",
+  "//third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.idl",
   "//third_party/blink/renderer/core/url/origin.idl",
   "//third_party/blink/renderer/core/url/url.idl",
   "//third_party/blink/renderer/core/url/url_search_params.idl",
diff --git a/third_party/blink/renderer/bindings/scripts/bind_gen/codegen_expr.py b/third_party/blink/renderer/bindings/scripts/bind_gen/codegen_expr.py
index aafb09af75d6adb990bf7b71f3ea5d4abe5da2dd..e7a7c0ec1e80b69c8cda9c474861fc770294c8a2 100644
--- a/third_party/blink/renderer/bindings/scripts/bind_gen/codegen_expr.py
+++ b/third_party/blink/renderer/bindings/scripts/bind_gen/codegen_expr.py
@@ -287,6 +287,7 @@ def expr_from_exposure(
         "DedicatedWorker": "IsDedicatedWorkerGlobalScope",
         "LayoutWorklet": "IsLayoutWorkletGlobalScope",
         "PaintWorklet": "IsPaintWorkletGlobalScope",
+        "PreloadRealm": "IsPreloadRealmGlobalScope",
         "ServiceWorker": "IsServiceWorkerGlobalScope",
         "ShadowRealm": "IsShadowRealmGlobalScope",
         "SharedWorker": "IsSharedWorkerGlobalScope",
diff --git a/third_party/blink/renderer/core/BUILD.gn b/third_party/blink/renderer/core/BUILD.gn
index 12942929d97d2bfff86fc43c506253934fd8e68e..a56872b923bd6944a9821ee455654b18daed26f1 100644
--- a/third_party/blink/renderer/core/BUILD.gn
+++ b/third_party/blink/renderer/core/BUILD.gn
@@ -65,6 +65,7 @@ import("//third_party/blink/renderer/core/page/build.gni")
 import("//third_party/blink/renderer/core/paint/build.gni")
 import("//third_party/blink/renderer/core/permissions_policy/build.gni")
 import("//third_party/blink/renderer/core/preferences/build.gni")
+import("//third_party/blink/renderer/core/preload_realm/build.gni")
 import("//third_party/blink/renderer/core/resize_observer/build.gni")
 import("//third_party/blink/renderer/core/route_matching/build.gni")
 import("//third_party/blink/renderer/core/sanitizer/build.gni")
@@ -361,6 +362,7 @@ component("core") {
   sources +=
       rebase_path(blink_core_sources_xmlhttprequest, "", "xmlhttprequest")
   sources += rebase_path(blink_core_sources_preferences, "", "preferences")
+  sources += rebase_path(blink_core_sources_preload_realm, "", "preload_realm")
 
   configs -= core_config_remove
   configs += core_config_add
diff --git a/third_party/blink/renderer/core/dom/abort_controller.idl b/third_party/blink/renderer/core/dom/abort_controller.idl
index 563b583bb9d64b350e3fdb0969901e8c674dc47c..427efba591af98680cfe59fa3f2089085d431599 100644
--- a/third_party/blink/renderer/core/dom/abort_controller.idl
+++ b/third_party/blink/renderer/core/dom/abort_controller.idl
@@ -5,7 +5,7 @@
 // https://dom.spec.whatwg.org/#interface-abortcontroller
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface AbortController {
     [CallWith=ScriptState, Measure] constructor();
     [SameObject] readonly attribute AbortSignal signal;
diff --git a/third_party/blink/renderer/core/dom/abort_signal.idl b/third_party/blink/renderer/core/dom/abort_signal.idl
index d4ed5c6639d848a71f40d866cf53516873e05292..99620265ccb9a98ea8513f75b55ecdff4b9c596f 100644
--- a/third_party/blink/renderer/core/dom/abort_signal.idl
+++ b/third_party/blink/renderer/core/dom/abort_signal.idl
@@ -5,7 +5,7 @@
 // https://dom.spec.whatwg.org/#interface-AbortSignal
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface AbortSignal : EventTarget {
     [
         CallWith=ScriptState,
diff --git a/third_party/blink/renderer/core/dom/dom_exception.idl b/third_party/blink/renderer/core/dom/dom_exception.idl
index 6e829626ffac53dd947b599ea374016ff73a8557..60a2a94c51167952aa0e7159f61c2ffd1e638770 100644
--- a/third_party/blink/renderer/core/dom/dom_exception.idl
+++ b/third_party/blink/renderer/core/dom/dom_exception.idl
@@ -31,7 +31,7 @@
 // https://webidl.spec.whatwg.org/#es-DOMException
 
 [
-    Exposed=(Window,Worker,ShadowRealm), // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm), // TODO(crbug.com/41480387): This should be Exposed=*
     Serializable
 ] interface DOMException {
     constructor(optional DOMString message = "", optional DOMString name = "Error");
diff --git a/third_party/blink/renderer/core/dom/events/custom_event.idl b/third_party/blink/renderer/core/dom/events/custom_event.idl
index 200c6b8f30b122fcbbd5c76fecdd83350c81a9b7..6c002b630bda97a3571a5b16121bb620773e91b8 100644
--- a/third_party/blink/renderer/core/dom/events/custom_event.idl
+++ b/third_party/blink/renderer/core/dom/events/custom_event.idl
@@ -26,7 +26,7 @@
 // https://dom.spec.whatwg.org/#interface-customevent
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface CustomEvent : Event {
     [CallWith=ScriptState] constructor(DOMString type, optional CustomEventInit eventInitDict = {});
     [CallWith=ScriptState] readonly attribute any detail;
diff --git a/third_party/blink/renderer/core/dom/events/event.idl b/third_party/blink/renderer/core/dom/events/event.idl
index 11e459da0f5d2b05d32ca8d483ed784473beed69..6cc8da25f79ca2aad3c52b79e17fd7b9a2e6280a 100644
--- a/third_party/blink/renderer/core/dom/events/event.idl
+++ b/third_party/blink/renderer/core/dom/events/event.idl
@@ -21,7 +21,7 @@
 // https://dom.spec.whatwg.org/#interface-event
 
 [
-    Exposed=(Window,Worker,AudioWorklet,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,AudioWorklet,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface Event {
     constructor(DOMString type, optional EventInit eventInitDict = {});
     readonly attribute DOMString type;
diff --git a/third_party/blink/renderer/core/dom/events/event_target.idl b/third_party/blink/renderer/core/dom/events/event_target.idl
index 24e54a9c02041b53c79657e8143f4f063f9cd161..803d937fabf31619dea6caa58b5e16c51fda0920 100644
--- a/third_party/blink/renderer/core/dom/events/event_target.idl
+++ b/third_party/blink/renderer/core/dom/events/event_target.idl
@@ -26,7 +26,7 @@ dictionary ObservableEventListenerOptions {
 };
 
 [
-    Exposed=(Window,Worker,AudioWorklet,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,AudioWorklet,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface EventTarget {
     [CallWith=ScriptState] constructor();
     undefined addEventListener(DOMString type, EventListener? listener, optional (AddEventListenerOptions or boolean) options);
diff --git a/third_party/blink/renderer/core/dom/observable.idl b/third_party/blink/renderer/core/dom/observable.idl
index 79410b5072c589f34cf180b10d1861e1a5516627..69cd9ce2fac2fe0f6cc1570ccd7a48bf8f1875c3 100644
--- a/third_party/blink/renderer/core/dom/observable.idl
+++ b/third_party/blink/renderer/core/dom/observable.idl
@@ -45,7 +45,7 @@ dictionary SubscribeOptions {
 typedef (ObserverCallback or Observer) ObserverUnion;
 typedef (ObserverCallback or ObservableInspector) ObservableInspectorUnion;
 
-[Exposed=(Window,Worker,ShadowRealm)] // TODO(crbug.com/40282760): This should be Exposed=*
+[Exposed=(Window,Worker,ShadowRealm,PreloadRealm)] // TODO(crbug.com/40282760): This should be Exposed=*
 interface Observable {
   [CallWith=ScriptState, MeasureAs=ObservableConstructor] constructor(SubscribeCallback callback);
   [CallWith=ScriptState] void subscribe(optional ObserverUnion observer = {}, optional SubscribeOptions options = {});
diff --git a/third_party/blink/renderer/core/dom/subscriber.idl b/third_party/blink/renderer/core/dom/subscriber.idl
index 9a09310ba30b872bc8b9d6f78d34d07717f61611..45dc4b90d9e818c36f4553730578ac5a00f497da 100644
--- a/third_party/blink/renderer/core/dom/subscriber.idl
+++ b/third_party/blink/renderer/core/dom/subscriber.idl
@@ -4,7 +4,7 @@
 
 // https://github.com/WICG/observable
 
-[Exposed=(Window,Worker,ShadowRealm)] // TODO(crbug.com/40282760): This should be Exposed=*
+[Exposed=(Window,Worker,ShadowRealm,PreloadRealm)] // TODO(crbug.com/40282760): This should be Exposed=*
 interface Subscriber {
   void next(any result);
   [CallWith=ScriptState] void error(any error);
diff --git a/third_party/blink/renderer/core/events/error_event.idl b/third_party/blink/renderer/core/events/error_event.idl
index cacba4123e3b97e085c3ea6023eb2ef3efc5a5b5..f176f64f7d3d323a83a5fb1a17a7d150ba032ca2 100644
--- a/third_party/blink/renderer/core/events/error_event.idl
+++ b/third_party/blink/renderer/core/events/error_event.idl
@@ -31,7 +31,7 @@
 // https://html.spec.whatwg.org/C/#the-errorevent-interface
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface ErrorEvent : Event {
     [CallWith=ScriptState] constructor(DOMString type, optional ErrorEventInit eventInitDict = {});
     readonly attribute DOMString message;
diff --git a/third_party/blink/renderer/core/events/promise_rejection_event.idl b/third_party/blink/renderer/core/events/promise_rejection_event.idl
index 693b2e78ee0e9dfb68bd0b08565ea76ce6711620..504f48fb2e3816686b75bb14798e51e722b62ffe 100644
--- a/third_party/blink/renderer/core/events/promise_rejection_event.idl
+++ b/third_party/blink/renderer/core/events/promise_rejection_event.idl
@@ -5,7 +5,7 @@
 // https://html.spec.whatwg.org/C/#promiserejectionevent
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface PromiseRejectionEvent : Event {
     [CallWith=ScriptState] constructor(DOMString type, PromiseRejectionEventInit eventInitDict);
     [CallWith=ScriptState] readonly attribute object promise;
diff --git a/third_party/blink/renderer/core/execution_context/execution_context.h b/third_party/blink/renderer/core/execution_context/execution_context.h
index 6c16b298502d3711ada242ad5d05aac10f1397a1..fed787a8c5bb16cd25055dfbbe89817eba71c163 100644
--- a/third_party/blink/renderer/core/execution_context/execution_context.h
+++ b/third_party/blink/renderer/core/execution_context/execution_context.h
@@ -160,6 +160,7 @@ class CORE_EXPORT ExecutionContext : public Supplementable<ExecutionContext>,
   virtual bool IsAudioWorkletGlobalScope() const { return false; }
   virtual bool IsLayoutWorkletGlobalScope() const { return false; }
   virtual bool IsPaintWorkletGlobalScope() const { return false; }
+  virtual bool IsPreloadRealmGlobalScope() const { return false; }
   virtual bool IsThreadedWorkletGlobalScope() const { return false; }
   virtual bool IsShadowRealmGlobalScope() const { return false; }
   virtual bool IsJSExecutionForbidden() const { return false; }
diff --git a/third_party/blink/renderer/core/inspector/worker_thread_debugger.cc b/third_party/blink/renderer/core/inspector/worker_thread_debugger.cc
index 3a52d94a105bcc34c2319940eb10ff381fe975ba..b3e3b1088239342d31090258e2d4b81d58d76fd0 100644
--- a/third_party/blink/renderer/core/inspector/worker_thread_debugger.cc
+++ b/third_party/blink/renderer/core/inspector/worker_thread_debugger.cc
@@ -38,6 +38,7 @@
 #include "third_party/blink/renderer/core/inspector/identifiers_factory.h"
 #include "third_party/blink/renderer/core/inspector/v8_inspector_string.h"
 #include "third_party/blink/renderer/core/inspector/worker_inspector_controller.h"
+#include "third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.h"
 #include "third_party/blink/renderer/core/shadow_realm/shadow_realm_global_scope.h"
 #include "third_party/blink/renderer/core/workers/dedicated_worker_global_scope.h"
 #include "third_party/blink/renderer/core/workers/worker_global_scope.h"
@@ -81,6 +82,10 @@ void WorkerThreadDebugger::ReportConsoleMessage(
     SourceLocation* location) {
   if (!context)
     return;
+  // An Electron preload realm has no worker thread of its own.
+  if (auto* preload_realm = DynamicTo<PreloadRealmGlobalScope>(context)) {
+    context = preload_realm->GetInitiatorExecutionContext();
+  }
 
   ExecutionContext* root_worker_context =
       context->IsShadowRealmGlobalScope()
@@ -168,6 +173,10 @@ void WorkerThreadDebugger::ExceptionThrown(WorkerThread* worker_thread,
 }
 
 int WorkerThreadDebugger::ContextGroupId(ExecutionContext* context) {
+  // An Electron preload realm belongs to its initiator's context group.
+  if (auto* preload_realm = DynamicTo<PreloadRealmGlobalScope>(context)) {
+    context = preload_realm->GetInitiatorExecutionContext();
+  }
   return ContextGroupId(To<WorkerOrWorkletGlobalScope>(context)->GetThread());
 }
 
diff --git a/third_party/blink/renderer/core/preload_realm/build.gni b/third_party/blink/renderer/core/preload_realm/build.gni
new file mode 100644
index 0000000000000000000000000000000000000000..95ad9051f0e81374443f4a7bb9d8e74fd39d230b
--- /dev/null
+++ b/third_party/blink/renderer/core/preload_realm/build.gni
@@ -0,0 +1,8 @@
+# Copyright (c) 2026 Anthropic, PBC.
+# Use of this source code is governed by the MIT license that can be
+# found in the LICENSE file.
+
+blink_core_sources_preload_realm = [
+  "preload_realm_global_scope.cc",
+  "preload_realm_global_scope.h",
+]
diff --git a/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.cc b/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.cc
new file mode 100644
index 0000000000000000000000000000000000000000..51a32a4a8d16c7286cd6d1c288397d6b93fc1454
--- /dev/null
+++ b/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.cc
@@ -0,0 +1,149 @@
+// Copyright (c) 2026 Anthropic, PBC.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.h"
+
+#include "base/notreached.h"
+#include "base/task/single_thread_task_runner.h"
+#include "services/metrics/public/cpp/ukm_source_id.h"
+#include "third_party/blink/public/platform/browser_interface_broker_proxy.h"
+#include "third_party/blink/renderer/core/inspector/console_message.h"
+#include "third_party/blink/renderer/core/inspector/inspector_audits_issue.h"
+
+namespace blink {
+
+PreloadRealmGlobalScope::PreloadRealmGlobalScope(
+    ExecutionContext* initiator_execution_context)
+    : ExecutionContext(initiator_execution_context->GetIsolate(),
+                       initiator_execution_context->GetAgent()),
+      initiator_execution_context_(initiator_execution_context) {}
+
+void PreloadRealmGlobalScope::Trace(Visitor* visitor) const {
+  visitor->Trace(initiator_execution_context_);
+  ScriptWrappable::Trace(visitor);
+  UniversalGlobalScope::Trace(visitor);
+  ExecutionContext::Trace(visitor);
+}
+
+ExecutionContext* PreloadRealmGlobalScope::GetExecutionContext() const {
+  return const_cast<PreloadRealmGlobalScope*>(this);
+}
+
+// Preload scripts reach the browser through Electron's IPC, never through
+// Blink's own interfaces.
+const BrowserInterfaceBrokerProxy&
+PreloadRealmGlobalScope::GetBrowserInterfaceBroker() const {
+  return GetEmptyBrowserInterfaceBroker();
+}
+
+scoped_refptr<base::SingleThreadTaskRunner>
+PreloadRealmGlobalScope::GetTaskRunner(TaskType task_type) {
+  return initiator_execution_context_->GetTaskRunner(task_type);
+}
+
+// Preload scripts are app code, so their feature usage is not counted as the
+// initiator's.
+void PreloadRealmGlobalScope::CountUse(mojom::blink::WebFeature feature) {}
+
+void PreloadRealmGlobalScope::CountDeprecation(
+    mojom::blink::WebFeature feature) {}
+
+void PreloadRealmGlobalScope::CountWebDXFeature(
+    mojom::blink::WebDXFeature feature) {}
+
+bool PreloadRealmGlobalScope::IsPreloadRealmGlobalScope() const {
+  return true;
+}
+
+const KURL& PreloadRealmGlobalScope::Url() const {
+  return initiator_execution_context_->Url();
+}
+
+const KURL& PreloadRealmGlobalScope::BaseURL() const {
+  return initiator_execution_context_->BaseURL();
+}
+
+KURL PreloadRealmGlobalScope::CompleteURL(const String& url) const {
+  return initiator_execution_context_->CompleteURL(url);
+}
+
+// No Content Security Policy is ever bound to a preload realm, so nothing
+// can ask it to disable eval.
+void PreloadRealmGlobalScope::DisableEval(const String& error_message) {
+  NOTREACHED();
+}
+
+void PreloadRealmGlobalScope::SetWasmEvalErrorMessage(
+    const String& error_message) {
+  NOTREACHED();
+}
+
+String PreloadRealmGlobalScope::UserAgent() const {
+  return initiator_execution_context_->UserAgent();
+}
+
+HttpsState PreloadRealmGlobalScope::GetHttpsState() const {
+  return initiator_execution_context_->GetHttpsState();
+}
+
+// No interface that loads resources is exposed to a preload realm.
+ResourceFetcher* PreloadRealmGlobalScope::Fetcher() {
+  NOTREACHED();
+}
+
+void PreloadRealmGlobalScope::ExceptionThrown(
+    ErrorEvent* error_event,
+    SanitizeScriptErrors sanitize_script_errors) {
+  initiator_execution_context_->ExceptionThrown(error_event,
+                                                sanitize_script_errors);
+}
+
+void PreloadRealmGlobalScope::AddInspectorIssue(AuditsIssue issue) {
+  initiator_execution_context_->AddInspectorIssue(std::move(issue));
+}
+
+// A preload realm has no `onerror`: uncaught errors go straight to
+// ExceptionThrown().
+EventTarget* PreloadRealmGlobalScope::ErrorEventTarget() {
+  return nullptr;
+}
+
+FrameOrWorkerScheduler* PreloadRealmGlobalScope::GetScheduler() {
+  return initiator_execution_context_->GetScheduler();
+}
+
+bool PreloadRealmGlobalScope::CrossOriginIsolatedCapability() const {
+  return false;
+}
+
+bool PreloadRealmGlobalScope::IsIsolatedContext() const {
+  return false;
+}
+
+ukm::UkmRecorder* PreloadRealmGlobalScope::UkmRecorder() {
+  return nullptr;
+}
+
+ukm::SourceId PreloadRealmGlobalScope::UkmSourceID() const {
+  return ukm::kInvalidSourceId;
+}
+
+// A preload realm identifies as its initiator, like an isolated world
+// identifies as its frame. A token type of its own would have to be taught to
+// every consumer of ExecutionContextToken in the browser process.
+ExecutionContextToken PreloadRealmGlobalScope::GetExecutionContextToken()
+    const {
+  return initiator_execution_context_->GetExecutionContextToken();
+}
+
+bool PreloadRealmGlobalScope::IsSecureContext() const {
+  return initiator_execution_context_->IsSecureContext();
+}
+
+void PreloadRealmGlobalScope::AddConsoleMessageImpl(ConsoleMessage* message,
+                                                    bool discard_duplicates) {
+  initiator_execution_context_->AddConsoleMessage(message, discard_duplicates);
+}
+
+}  // namespace blink
diff --git a/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.h b/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.h
new file mode 100644
index 0000000000000000000000000000000000000000..4d525b9ddf3009dd93280901a129a72c7a8a8990
--- /dev/null
+++ b/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.h
@@ -0,0 +1,89 @@
+// Copyright (c) 2026 Anthropic, PBC.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#ifndef THIRD_PARTY_BLINK_RENDERER_CORE_PRELOAD_REALM_PRELOAD_REALM_GLOBAL_SCOPE_H_
+#define THIRD_PARTY_BLINK_RENDERER_CORE_PRELOAD_REALM_PRELOAD_REALM_GLOBAL_SCOPE_H_
+
+#include "base/task/single_thread_task_runner.h"
+#include "third_party/blink/renderer/core/core_export.h"
+#include "third_party/blink/renderer/core/execution_context/execution_context.h"
+#include "third_party/blink/renderer/core/frame/universal_global_scope.h"
+#include "third_party/blink/renderer/platform/bindings/script_wrappable.h"
+#include "third_party/blink/renderer/platform/weborigin/kurl.h"
+#include "third_party/blink/renderer/platform/wtf/casting.h"
+
+namespace blink {
+
+// Global scope of the context Electron runs preload scripts in when the
+// context they are attached to (the "initiator") has no isolated worlds of its
+// own, as is the case for service workers. It exposes only the interfaces
+// marked [Exposed=PreloadRealm] or [Exposed=*] and otherwise defers to the
+// initiator, which owns it.
+class CORE_EXPORT PreloadRealmGlobalScope final : public ScriptWrappable,
+                                                  public UniversalGlobalScope,
+                                                  public ExecutionContext {
+  DEFINE_WRAPPERTYPEINFO();
+
+ public:
+  explicit PreloadRealmGlobalScope(
+      ExecutionContext* initiator_execution_context);
+
+  ExecutionContext* GetInitiatorExecutionContext() const {
+    return initiator_execution_context_.Get();
+  }
+
+  void Trace(Visitor* visitor) const override;
+
+  // UniversalGlobalScope:
+  ExecutionContext* GetExecutionContext() const override;
+
+  // MojoBindingContext:
+  const BrowserInterfaceBrokerProxy& GetBrowserInterfaceBroker() const override;
+  scoped_refptr<base::SingleThreadTaskRunner> GetTaskRunner(
+      TaskType task_type) override;
+
+  // UseCounter:
+  void CountUse(mojom::blink::WebFeature feature) override;
+  void CountDeprecation(mojom::blink::WebFeature feature) override;
+  void CountWebDXFeature(mojom::blink::WebDXFeature feature) override;
+
+  // ExecutionContext:
+  bool IsPreloadRealmGlobalScope() const override;
+  const KURL& Url() const override;
+  const KURL& BaseURL() const override;
+  KURL CompleteURL(const String& url) const override;
+  void DisableEval(const String& error_message) override;
+  void SetWasmEvalErrorMessage(const String& error_message) override;
+  String UserAgent() const override;
+  HttpsState GetHttpsState() const override;
+  ResourceFetcher* Fetcher() override;
+  void ExceptionThrown(ErrorEvent* error_event,
+                       SanitizeScriptErrors sanitize_script_errors) override;
+  void AddInspectorIssue(AuditsIssue issue) override;
+  EventTarget* ErrorEventTarget() override;
+  FrameOrWorkerScheduler* GetScheduler() override;
+  bool CrossOriginIsolatedCapability() const override;
+  bool IsIsolatedContext() const override;
+  ukm::UkmRecorder* UkmRecorder() override;
+  ukm::SourceId UkmSourceID() const override;
+  ExecutionContextToken GetExecutionContextToken() const override;
+  bool IsSecureContext() const override;
+
+ private:
+  void AddConsoleMessageImpl(ConsoleMessage* message,
+                             bool discard_duplicates) override;
+
+  const Member<ExecutionContext> initiator_execution_context_;
+};
+
+template <>
+struct DowncastTraits<PreloadRealmGlobalScope> {
+  static bool AllowFrom(const ExecutionContext& context) {
+    return context.IsPreloadRealmGlobalScope();
+  }
+};
+
+}  // namespace blink
+
+#endif  // THIRD_PARTY_BLINK_RENDERER_CORE_PRELOAD_REALM_PRELOAD_REALM_GLOBAL_SCOPE_H_
diff --git a/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.idl b/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.idl
new file mode 100644
index 0000000000000000000000000000000000000000..c74c96d418c86996ca4754f54751e15591478357
--- /dev/null
+++ b/third_party/blink/renderer/core/preload_realm/preload_realm_global_scope.idl
@@ -0,0 +1,12 @@
+// Copyright (c) 2026 Anthropic, PBC.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+// Global scope of the context Electron runs service worker preload scripts in.
+
+[
+    Exposed=PreloadRealm,
+    Global=PreloadRealm
+] interface PreloadRealmGlobalScope {};
+
+PreloadRealmGlobalScope includes UniversalGlobalScope;
diff --git a/third_party/blink/renderer/core/url/url.idl b/third_party/blink/renderer/core/url/url.idl
index 2f83c1225068ac03d3ebf75515cf0f4ea191db2d..74abf60de4f4922f168abd5b055fb1ef2a869cf2 100644
--- a/third_party/blink/renderer/core/url/url.idl
+++ b/third_party/blink/renderer/core/url/url.idl
@@ -27,7 +27,7 @@
 // https://url.spec.whatwg.org/#url
 
 [
-    Exposed=(Window,Worker,ShadowRealm), // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm), // TODO(crbug.com/41480387): This should be Exposed=*
     LegacyWindowAlias=webkitURL
 ] interface URL {
     [RaisesException] constructor(USVString url, optional USVString base);
diff --git a/third_party/blink/renderer/core/url/url_search_params.idl b/third_party/blink/renderer/core/url/url_search_params.idl
index 2ac09cbcf12b2bb0213f0a117d8981ad212f9719..ceb2903ae6a7d084db4e1b0f5e002a41f8b3c2cb 100644
--- a/third_party/blink/renderer/core/url/url_search_params.idl
+++ b/third_party/blink/renderer/core/url/url_search_params.idl
@@ -5,7 +5,7 @@
 // https://url.spec.whatwg.org/#interface-urlsearchparams
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface URLSearchParams {
     [RaisesException] constructor(optional (sequence<sequence<USVString>> or record<USVString, USVString> or USVString) init = "");
     readonly attribute unsigned long size;
diff --git a/third_party/blink/renderer/core/url_pattern/url_pattern.idl b/third_party/blink/renderer/core/url_pattern/url_pattern.idl
index e82a9bb83d39f69191077299144c2b5508274f03..9b44dcb473ee847951e8a05e91c87bc19d869aae 100644
--- a/third_party/blink/renderer/core/url_pattern/url_pattern.idl
+++ b/third_party/blink/renderer/core/url_pattern/url_pattern.idl
@@ -10,7 +10,7 @@ enum URLPatternComponent { "protocol", "username", "password", "hostname",
 
 // https://wicg.github.io/urlpattern/
 [
-  Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+  Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface URLPattern {
   [RaisesException, CallWith=Isolate, Measure]
   constructor(URLPatternInput input, USVString baseURL, optional URLPatternOptions options = {});
diff --git a/third_party/blink/renderer/modules/compression/compression_stream.idl b/third_party/blink/renderer/modules/compression/compression_stream.idl
index 78652ab22ddfac35fa95f65dc5574e2b6ff20170..1243baa84f2fb88423070df12e26f057ef1c5d25 100644
--- a/third_party/blink/renderer/modules/compression/compression_stream.idl
+++ b/third_party/blink/renderer/modules/compression/compression_stream.idl
@@ -1,5 +1,5 @@
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface CompressionStream {
     [CallWith=ScriptState, RaisesException, MeasureAs=CompressionStreamConstructor] constructor(DOMString format);
     readonly attribute ReadableStream readable;
diff --git a/third_party/blink/renderer/modules/compression/decompression_stream.idl b/third_party/blink/renderer/modules/compression/decompression_stream.idl
index 7856196ec0ac3362b4745a22045f4ee71ef6370d..1c325435ccec3d29382fa394e784c131edccd83e 100644
--- a/third_party/blink/renderer/modules/compression/decompression_stream.idl
+++ b/third_party/blink/renderer/modules/compression/decompression_stream.idl
@@ -1,5 +1,5 @@
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface DecompressionStream {
     [CallWith=ScriptState, RaisesException, MeasureAs=DecompressionStreamConstructor] constructor(DOMString format);
     readonly attribute ReadableStream readable;
diff --git a/third_party/blink/renderer/modules/encoding/text_decoder.idl b/third_party/blink/renderer/modules/encoding/text_decoder.idl
index 4564e37db341e567fab74ddf93dd5f90423e3f4e..8f83f70b372d27e9176232dc7fe6c3af89fc44a1 100644
--- a/third_party/blink/renderer/modules/encoding/text_decoder.idl
+++ b/third_party/blink/renderer/modules/encoding/text_decoder.idl
@@ -31,7 +31,7 @@
 // https://encoding.spec.whatwg.org/#textdecoder
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface TextDecoder {
     [RaisesException, MeasureAs=TextDecoderConstructor] constructor(optional DOMString label = "utf-8", optional TextDecoderOptions options = {});
     readonly attribute DOMString encoding;
diff --git a/third_party/blink/renderer/modules/encoding/text_decoder_stream.idl b/third_party/blink/renderer/modules/encoding/text_decoder_stream.idl
index b5f09ee82ff68cd7dab3370fbefb42a96fc371df..09b2bf58080085d8eea8f46022cc97fb47ef7397 100644
--- a/third_party/blink/renderer/modules/encoding/text_decoder_stream.idl
+++ b/third_party/blink/renderer/modules/encoding/text_decoder_stream.idl
@@ -4,7 +4,7 @@
 
 // https://encoding.spec.whatwg.org/#interface-textdecoderstream
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface TextDecoderStream {
     [CallWith=ScriptState, RaisesException, MeasureAs=TextDecoderStreamConstructor] constructor(optional DOMString label = "utf-8", optional TextDecoderOptions options = {});
     readonly attribute DOMString encoding;
diff --git a/third_party/blink/renderer/modules/encoding/text_encoder.idl b/third_party/blink/renderer/modules/encoding/text_encoder.idl
index 8085e6edc8349a9d3763f825313cc3445bf06312..245d53014a80e61b24120ff898c6398b3ef3e07c 100644
--- a/third_party/blink/renderer/modules/encoding/text_encoder.idl
+++ b/third_party/blink/renderer/modules/encoding/text_encoder.idl
@@ -29,7 +29,7 @@
  */
 
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface TextEncoder {
     [CallWith=ExecutionContext, RaisesException, MeasureAs=TextEncoderConstructor] constructor();
     readonly attribute DOMString encoding;
diff --git a/third_party/blink/renderer/modules/encoding/text_encoder_stream.idl b/third_party/blink/renderer/modules/encoding/text_encoder_stream.idl
index d2c71dee48bbe0a08e4ba6dee7a28cdbef6c52a2..4daa2b3eca5da7a80475a722fb3c96b8ec3c3fc7 100644
--- a/third_party/blink/renderer/modules/encoding/text_encoder_stream.idl
+++ b/third_party/blink/renderer/modules/encoding/text_encoder_stream.idl
@@ -4,7 +4,7 @@
 
 // https://encoding.spec.whatwg.org/#interface-textencoderstream
 [
-    Exposed=(Window,Worker,ShadowRealm) // TODO(crbug.com/41480387): This should be Exposed=*
+    Exposed=(Window,Worker,ShadowRealm,PreloadRealm) // TODO(crbug.com/41480387): This should be Exposed=*
 ] interface TextEncoderStream {
     [CallWith=ScriptState, RaisesException, MeasureAs=TextEncoderStreamConstructor] constructor();
     readonly attribute DOMString encoding;
